feat: Connection String Support

matthewmcneely opened the pull request:
This PR adds support to open gRPC endpoints using "connection strings".

<h2><code>dgraph://{username:password@}host:port?args</code></h2>
<p><code>username</code> and <code>password</code> are optional. If username is provided, a password must also be present. If supplied, these credentials are used to log into the cluster thru the ACL mechanism.</p>
<p>Valid args:</p>

Arg | Value | Description
-- | -- | --
apikey | \<key\> | This is a Dgraph Cloud API key
bearertoken | \<access token\> | An access token
sslmode | disable | The default. Do not connect via TLS
  | require | Use TLS to connect
  | verify-ca | Use TLS and verify the certificate found


<p>Note, the sslmode options are borrowed directly from postgres</p>
<p>gRPC clients that encounter a dgraph connection string use the url to create a gRPC connection. In HTTP-based clients and tools that we control (Ratel, dgraph-js-http), connection code can check for the <code>dgraph://</code> prefix and rework into HTTP endpoints—although some (potentially incorrect) guessing on ports might be required when the target cluster is not controlled by us.</p>

**Checklist**

- [x] Code compiles correctly and linting passes locally
- [x] Tests added for new functionality, or regression tests for bug fixes added as applicable

matthewmcneely commented:
Getting closer. Still need to sort out testing against clusters running self-certificates. The tests are all @skipped for now. I'm working on a Dgraph cluster runner tool in the dgraph-experimental repo that would allow testing against all the different flavors of on-prem deployments (acl, self-cert, trusted-cert, etc)

mangalaman93 commented:
@matthewmcneely dgo repo should just have basic tests and as many unit tests as possible. We already have dgraphtest package in Dgraph repo that tests all scenarios.

matthewmcneely commented:
@mangalaman93 That (dgraphtest) seems to cover Cloud testing and local clusters with ACL, but not on-prem tls configurations like signed certs, etc. Also, looks like the that dgraph package (dgraphtest) will need to be upgraded to the new Open/connection string concept (or the NewClient func perhaps...

matthewmcneely commented on client.go:
```diff
@@ -54,13 +64,68 @@ func (a *authCreds) RequireTransportSecurity() bool {
+			return nil, fmt.Errorf("invalid API key")
+		}
+		if !certPoolAdded {
+			opts = append(opts, WithSystemCertPool())
```
> Our default is to use a TLS connection with system cert pool

It is? By default, the current dgo connects insecurely. We shouldn't change this default behaviour. In my mind, sslmode is 'disable' by default.

mangalaman93 replied:
I do not think current dgo has insecure by default. The current setup is completely up to the user as to how they set it up, we do not control that process today inside dgo. Hence, our default choice should be a secure option when we are responsible for setting up the connection unlike before.

matthewmcneely replied:
Right, but I'm thinking of the "out-of-the-box" Dgraph setup, just like postgres which does not have TLS setup by default. Making the default sslmode 'require' would force devs to dive into the docs to understand that they have to add the sslmode=disable param to their connection string.

Let me know as I'll need to rework Open completely to support the default sslmode of 'require'.

mangalaman93 replied:
okay, I think we can do default sslmode=disable. That way users with local Dgraph setup can use Open as well with a simple "localhost:8080" string.

matthewmcneely commented on client.go:
```diff
@@ -54,13 +64,68 @@ func (a *authCreds) RequireTransportSecurity() bool {
+
+	certPoolAdded := false
+	if sslMode, ok := u.Query()[sslModeParam]; ok {
+		if sslMode[0] == string(sslModeRequire) {
```
Right. I reworked the Open function completely. Note I added a new sslmode enum, `verify-ca` which verifies the server certificate is signed by a trusted CA (this is the same as postgres).

matthewmcneely commented:
Note, I've also removed the login_test module from this PR as it was really for my sanity testing. I've updated the test code to use .Open(). 

Once we arrive at a final code change set, I'll update the README with the new connection string options.

mangalaman93 commented on client.go:
```diff
@@ -54,13 +63,76 @@ func (a *authCreds) RequireTransportSecurity() bool {
+		return nil, fmt.Errorf("invalid connection string: %w", err)
+	}
+
+	if !strings.HasPrefix(connStr, dgraphScheme) {
```
can we not u.schema here instead?

mangalaman93 commented on client.go:
```diff
@@ -54,13 +63,76 @@ func (a *authCreds) RequireTransportSecurity() bool {
+	}
+
+	if !strings.HasPrefix(connStr, dgraphScheme) {
+		return nil, fmt.Errorf("invalid connection string: must start with %s", dgraphScheme)
```
we should say, invalid connection schema here

mangalaman93 replied:
what if I do not specify anything, can we not accept that as well? I am thinking that "localhost:8080" should work as well

matthewmcneely replied:
We could accept anything I guess. But localhost:8080 parses oddly. The scheme becomes the host and the host is nil...

```go
u, err := url.Parse("localhost:8080")
fmt.Println("scheme:", u.Scheme)
fmt.Println("host:", u.Host)
fmt.Println("error:", err)
```

```sh
scheme: localhost
host: 
error: <nil>
```

postgres doesn't support it. Slippery slope IMO.

mangalaman93 commented on client.go:
```diff
@@ -54,13 +63,76 @@ func (a *authCreds) RequireTransportSecurity() bool {
+	sslMode := u.Query().Get(sslModeParam)
+
+	if apiKey != "" {
+		if len(apiKey) < 12 {
```
I am not sure that we should assume this

mangalaman93 commented on client.go:
```diff
@@ -54,13 +63,76 @@ func (a *authCreds) RequireTransportSecurity() bool {
+		if len(apiKey) < 12 {
+			return nil, errors.New("invalid API key")
+		}
+		// Dgraph Cloud requires SSL
```
again, we should just follow what the connection string says, we should not do this by default.

mangalaman93 commented on README.md:
```diff
@@ -49,7 +49,8 @@ Note: One of the most important API breakages from dgo v1 to v2 is in the functi
 This function returns an `*api.Assigned` value in v1 but an `*api.Response` in v2.
```
I am thinking that we do not update the readme yet until we are in preview. Please create a separate PR for this change.

matthewmcneely replied:
So revert the entire README?

mangalaman93 replied:
just put in a separate PR that w3e are going to merge right before the release?

ryanfoxtyler replied:
I thought these were backward compatible–what breaks if we merge this now?

mangalaman93 commented on client.go:
```diff
@@ -54,13 +63,71 @@ func (a *authCreds) RequireTransportSecurity() bool {
+	apiKey := u.Query().Get(cloudAPIKeyParam)
+	sslMode := u.Query().Get(sslModeParam)
+
+	if apiKey != "" {
```
what if I provide a hypermode key?

matthewmcneely replied:
@ryanfoxtyler Need guidance here. Is the mechanism for accepting passed Bearer tokens in place in modus? If so, we'll need to add a new url param, e.g., moduskey or something

mangalaman93 replied:
We already have a function with name WithHypermodeKey or something

ryanfoxtyler replied:
Is the concern that Dgraph Cloud API keys do not use the `Bearer` prefix but Hypermode does (as per standard)?

matthewmcneely replied:
@ryanfoxtyler NM.... Aman and I have a solution for this. I'll notify you of the proposed solution when it's up.

mangalaman93 commented on client.go:
```diff
@@ -54,13 +63,71 @@ func (a *authCreds) RequireTransportSecurity() bool {
+		return nil, fmt.Errorf("invalid SSL mode: %s (must be one of %s, %s, %s)", sslMode, sslModeDisable, sslModeRequire, sslModeVerifyCA)
+	}
+
+	if u.User != nil {
```
we should check for both here. If only one of them is present, we should let the user know that something is wrong.

matthewmcneely replied:
Here u.User is an object containing both username and password

mangalaman93 replied:
but what if I have provided the username and not provided the password? Will I go inside this if then?

matthewmcneely commented on t/docker-compose-latest.yml:
```diff
@@ -0,0 +1,60 @@
+version: "3.5"
```
the *-latest is a compose file allowing testing against the docker image, not the mounted dgraph exec
